pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1332,79 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
 		cd.Annotations = map[string]string{}
 	}
 }
+
+// managePullSecrets() takes pull secretes in to account and returns the merged pull request
+func (r *ReconcileClusterDeployment) managePullSecrets(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (string, error) {
+
+	// For code readability let's call the pull secret in cluster deployment config as local pull secret
+	localPullSecret, err := controllerutils.LoadSecretData(r.Client, cd.Spec.PullSecret.Name, cd.Namespace, corev1.DockerConfigJsonKey)
+	if err != nil {
+		cdLog.Warn("unable to load pull secret from secret")
+	}
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+	if len(globalPullSecret) == 0 {
+		cdLog.Warn("global pull secret not available")
+		if localPullSecret == "" {
+			errMsg := "Pull request or global pull request not available"
+			cdLog.Error(errMsg)
+			return "", fmt.Errorf("%s", errMsg)
+		}
+	}
+
+	// If global pull secret is present then merge the pullSecret and globalPullSecret
+	if globalPullSecret != "" && localPullSecret != "" {
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", fmt.Errorf("%s : %v", errMsg, err)
+		}
+		return pullSecret, nil
+	} else if globalPullSecret != "" {
+		return globalPullSecret, nil
+	} else {
+		return localPullSecret, nil
+	}
+}
+
+// updatePullSecretObj updates pull secret object in cluster deployment status if required
+func (r *ReconcileClusterDeployment) updatePullSecretObj(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+
+	// Check if pull secret object already exists
+	existingPullSecretObj := r.getExistingPullSecret(cd)
+	pullSecretObj := generatePullSecretObj(pullSecret, cd)
+
+	var pullSecretHash, existingPullSecretHash string
+
+	if existingPullSecretObj != nil {
+		existingPullSecretHash = controllerutils.CalculateHashOfSecret(existingPullSecretObj)
```

Comment:
Why do we need to hash the entire `Secret`? We only care about the pull secret--which is the value of the specific key of the `Secret`. Otherwise, this is always doing to fail unless we copy the creation timestamp into the new secret.

Reply:
Here is an example of the merged pull secret. 
`expectedStr: `{"auths":{"cloud.openshift.com":{"auth":"b34xVjWERckjfUyV1pMQTc=","email":"lmohanty@redhat.com"},"quay.io":{"auth":"b3BlbnNoVkc=","email":"lmohanty@redhat.com"},"registry.connect.redhat.com":{"auth":"NjQ4ODeDZ3d1pN","email":"lmohanty@redhat.com"},"registry.redhat.io":{"auth":"NjQ4ODX1pN","email":"lmohanty@redhat.com"},"registry.svc.ci.openshift.org":{"auth":"dXNljlfjldsfSDD"}}}`,`

Reply:
Ah! you mean just retrieve the pull secret and findout the has rather than the secret. Sure we can do that. But does it make any difference as the secret is not a big structure either?

Reply:
I think @staebler meant to hash or compare the one specific key in secret.Data[], as the metadata will always show a diff here otherwise.

Reply:
We already have https://github.com/openshift/hive/blob/master/pkg/controller/remoteingress/remoteingress_controller.go#L437-L456 , so I am going to reuse that. It also inline to your suggestion.

Reply:
You *can* reuse that code. It is a bit of extra work to build up a `Secret` object just to use that function. I think that all you need to do is the following. We only care about the one key in the secret, so there is no need to create a map of the entire `Secret` data.
```
fmt.Sprintf("%x", md5.Sum(pullSecret))
```

Reply:
We should reuse code unless it is not efficient. We need to the code anyway to build the object and we have code to create shasum. So IMO we should just reuse.

Reply:
One thing I noticed in https://github.com/openshift/hive/blob/master/pkg/controller/remoteingress/remoteingress_controller.go#L437-L456 is that it can be simplified using `json.Marshal()` to serialize the data. Will look in to this.

Reply:
I think that you will find that that function does not work for your purposes. In order to use that function, you will need to base64-encode the pull secret and set it in `Data` instead of setting `StringData`.

Reply:
This is resolved now. Let me know if you find otherwise.